python/private/py_wheel.bzl

```diff
@@ -203,6 +203,427 @@ _DESCRIPTION_FILE_EXTENSION_TO_TYPE = {
+            elif escaped.endswith("_"):
+                pass
+            else:
+                escaped += "_"
+        else:
+            fail(
+                "A valid name consists only of ASCII letters ",
+                "and numbers, period, underscore and hyphen.",
+                "Name '%s' has bad character '%s'." % (name, character),
+            )
+    if escaped.endswith("_"):
+        fail(
+            "A valid name must end with a letter or number.",
+            "Name '%s' does not." % name,
+        )
+    return escaped
+
+def normalize_pep440(version):
+    """Escape the version component of a filename.
+
+    See https://packaging.python.org/en/latest/specifications/binary-distribution-format/#escaping-and-unicode
+    and https://peps.python.org/pep-0440/
+
+    Args:
+      version: version string to be normalized according to PEP 440.
+
+    Returns:
+      string containing the normalized version.
+    """
+
+    version = version.strip()  # PEP 440: Leading and Trailing Whitespace
+    contexts = []
+
+    def open_context(start):
+        """Open an new parsing context.
+
+        If the current parsing step succeeds, call close_context().
+        If the current parsing step fails, call contexts.pop() to
+        go back to how it was before we opened a new context.
+
+        Args:
+          start: index into `version` indicating where the current
+            parsing step starts.
+        """
+        contexts.append({"norm": "", "start": start})
+        return contexts[-1]
+
+    def close_context():
+        """Close the current context successfully and merge the results."""
+        finished = contexts.pop()
+        contexts[-1]["norm"] += finished["norm"]
+        contexts[-1]["start"] = finished["start"]
+
+    def is_(reference):
+        """Predicate testing a token for equality with `reference`."""
+        return lambda token: token == reference
+
+    def is_not(reference):
+        """Predicate testing a token for inequality with `reference`."""
+        return lambda token: token != reference
+
+    def in_(reference):
+        """Predicate testing if a token is in the list `reference`."""
+        return lambda token: token in reference
+
+    def accept(predicate, value):
+        """If `predicate` matches the next token, accept the token.
+
+        Accepting the token means adding it (according to `value`) to
+        the running results maintained in context["norm"] and
+        advancing the cursor in context["start"] to the next token in
+        `version`.
+
+        Args:
+          predicate: function taking a token and returning a boolean
+            saying if we want to accept the token.
+          value: the string to add if there's a match, or, if `value`
+            is a function, the function to apply to the current token
+            to get the string to add.
+
+        Returns:
+          whether a token was accepted.
+        """
+
+        context = contexts[-1]
```

Comment:
consider renaming the variable to `ctx` and making a `parser.context()` function to access the last context without accessing a private `contexts` array. Right now `contexts` implements a stack, but it could also in theory be a linked list and the interface of the parser should be agnostic to the underlying data structure.

Reply:
Done.